### PR TITLE
types:check correctness of decimal's exponent part

### PR DESCRIPTION
**Pull request by lysu**
### fix int overflow

```
drop table if exists t1;
create table `t1` (`a` decimal(10,2));
insert into t1 values ("1e+18446744073709551616")
```

expect:
```
ERROR 1366 (HY000): Incorrect decimal value: '1e+18446744073709551616' for column 'a' at row 1
```

but got:
```
Query OK, 1 row affected (0.01 sec)

mysql> select * from t1;
+------+
| a    |
+------+
| 1.00 |
+------+
```
The issue is cause by exponent part overflow int max.

### fix empty exponent

in this issue we also will fix

```
insert into t1 values ("123.4e")
```

expect:
```
ERROR 1366 (HY000): Incorrect decimal value: '123.4e' for column 'a' at row 1
```

but got
```
Query OK, 1 row affected, 1 warning (0.01 sec)
```

### CODE REVIEW

File: types/mydecimal.go [hunk 1]
```diff
@@ -391,9 +391,8 @@ func (d *MyDecimal) FromString(str []byte) error {
 	if innerIdx != 0 {
 		d.wordBuf[wordIdx] = word * powers10[digitsPerWord-innerIdx]
 	}
-	if endIdx+1 < len(str) && (str[endIdx] == 'e' || str[endIdx] == 'E') {
-		exponent, err1 := strToInt(string(str[endIdx+1:]))
-		// TODO: need a way to check if there is at least one digit.
+	if endIdx+1 <= len(str) && (str[endIdx] == 'e' || str[endIdx] == 'E') {
+		exponent, err1 := strconv.ParseInt(string(str[endIdx+1:]), 10, 64)
```
Comment (coocood) [inline types/mydecimal.go]: If the exponent part is `e2e`, then we the exponent should be `2`, instead of `0`.

File: types/mydecimal_test.go [hunk 2]
```diff
@@ -380,11 +380,16 @@ func (s *testMyDecimalSuite) TestFromString(c *C) {
 		{"1234500009876.5", "1234500009876.5", nil},
 		{"123E5", "12300000", nil},
 		{"123E-2", "1.23", nil},
+		{"1e18446744073709551615", "0", ErrBadNumber},
+		{"1e", "0", ErrBadNumber},
```
Comment (coocood) [inline types/mydecimal_test.go]: Please add a case `1e1e`.

File: types/mydecimal_test.go [hunk 3]
```diff
@@ -380,11 +380,16 @@ func (s *testMyDecimalSuite) TestFromString(c *C) {
...
 	}
 	for _, ca := range tests {
 		var dec MyDecimal
 		err := dec.FromString([]byte(ca.input))
 		c.Check(err, Equals, ca.err)
+		if err != nil {
+			continue
```
Comment (zhexuany) [inline types/mydecimal_test.go]: Why do we ignore error here?

File: types/mydecimal_test.go [hunk 4]
```diff
@@ -380,13 +380,22 @@ func (s *testMyDecimalSuite) TestFromString(c *C) {
...
+		{"1e18446744073709551615", "0", ErrOverflow},
+		{"1e213123213123", "0", ErrOverflow},
+		{"1e1073741823", "0", ErrOverflow},
+		{"1e18446744073709551620", "0", ErrOverflow},
+		{"1e", "1", ErrBadNumber},
+		{"1e1e", "10", ErrBadNumber},
+		{"1e001", "10", nil},
+		{"1e00", "1", nil},
+		{"1eabc", "1", ErrBadNumber},
+		{"1e 1dddd ", "10", ErrBadNumber},
```
Comment (coocood) [inline types/mydecimal_test.go]: The result should be "1"

**Comment by lysu** [comment]
# TiDB Decimal overflow action after this PR

it will compatible with MySQL.

- exponent not a valid number will cause ERROR when sql_mode has `STRICT_TRANS_TABLES`
- it will try its best to got a value when sql_mode without `STRICT_TRANS_TABLES` and add warning info into `show warnings`

For SQL_MODE without `STRICT_TRANS_TABLES` situtation, there are more sub-situtation need consider.

- space at both end will be trimed.
- if exponent not all number(eg. `1dbc`), it will read&parse str and stop at first character that is not a digit(`11abc` will be `11`), then return read number `n` as exponent & add `Data Truncated` warn. final value is `xe0` as `x` with data truncated warn.
- if exponent part is empty or no any number will return `0` as exponent & add `Data Truncated` warn, final value is `xe0` with data truncated warn.
- if exponent part over int64 max/min will return MAXINT32/MININT32 & add `Bad Number` warn, final value is `0` with bad number warn.
- if exponent part valid, but final value overflow max/min, final value will be `99999999.99` max decimal with `out of range warn`

**Comment by lysu** [comment]
It's PR only make sure generate right internal error, but there still some question.

- ErrDataTruncate for final user is miss column info
- Out of range error for final user is wrong code and wrong error message
- ErrBadNumber ditto

those question will be care in #6610

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@lysu 
Can we remove the WIP  tag? since the remained issue is handled by another PR.

**Comment by lysu** [comment]
@shenli @zhexuany @tiancaiamao PTLA this RP, thx.

File: types/helper.go [hunk 5]
```diff
@@ -143,16 +148,47 @@ func strToInt(str string) (int64, error) {
...
+	r := uint64(0)
 	for ; i < len(str); i++ {
 		if !unicode.IsDigit(rune(str[i])) {
+			err = ErrTruncated
+			break
+		}
+		hasNum = true
+		if r >= uintCutOff {
+			r = 0
+			err = ErrBadNumber
```
Comment (tiancaiamao) [inline types/helper.go]: errors.Trace(err)

File: types/helper.go [hunk 6]
```diff
@@ -143,16 +148,47 @@ func strToInt(str string) (int64, error) {
...
+			break
+		}
+		r = r * uint64(10)
+
+		r1 := r + uint64(str[i]-'0')
+		if r1 < r || r1 > maxUint {
+			r = 0
+			err = ErrBadNumber
```
Comment (tiancaiamao) [inline types/helper.go]: ditto

File: types/mydecimal.go [hunk 7]
```diff
@@ -391,23 +391,30 @@ func (d *MyDecimal) FromString(str []byte) error {
...
 	}
-	if endIdx+1 < len(str) && (str[endIdx] == 'e' || str[endIdx] == 'E') {
+	if endIdx+1 <= len(str) && (str[endIdx] == 'e' || str[endIdx] == 'E') {
 		exponent, err1 := strToInt(string(str[endIdx+1:]))
-		// TODO: need a way to check if there is at least one digit.
 		if err1 != nil {
-			*d = zeroMyDecimal
-			return ErrBadNumber
+			err = err1
+			if err != ErrTruncated {
```
Comment (tiancaiamao) [inline types/mydecimal.go]: errors.Cause(err)

File: types/helper.go (hunk 5 above)
Comment (tiancaiamao) [reply types/helper.go]: If there are multiple error, say ErrBadNumber, in one function, we can't trace which is the actually cause of the error.
Do you remember I've ever write a error "invalid time format" without detail stack trace? And fix that introduce another PR, which could have been avoid. @coocood

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**